### PR TITLE
Prevent unrealistic eager memory alloc in conv1x1 (my bad)

### PR DESCRIPTION
**Pull request by mloubout**
@rafaelorozco 

Update compat and dependencies versions
Cleanup activations (better perf too)
Refactor base abstract and concrete types for slightly easier codebase

@grizzuti the "generalization" changed the ordering of parameter retrieval. I don't think it's an issue but may be a problem for older version saved parameters to be loaded.

### CODE REVIEW

**Comment by mloubout** [comment]
@rafaelorozco from the test log (bottm of it) seems slightly better for memory now. Looking into IRIM part and the cat/...

**Comment by alisiahkoohi** [comment]
> @rafaelorozco
> 
> Update compat and dependencies versions Cleanup activations (better perf too) Refactor base abstract and concrete types for slightly easier codebase
> 
> @grizzuti the "generalization" changed the ordering of parameter retrieval. I don't think it's an issue but may be a problem for older version saved parameters to be loaded.

Would it be possible not to change the order of parameter retrieval? As you said, may cause issue older pretrained models.

**Comment by mloubout** [comment]
>Would it be possible not to change the order of parameter retrieval? As you said, may cause issue older pretrained models.

Is kinda of a pain because right now it's done by hand on each network but there is not easy "generalized" way to get the order obtained by hand because depending on the network it makes and implicit  assumption on the order in which the layers (struct's field) are used.

File: src/layers/invertible_layer_actnorm.jl [hunk 1]
```diff
@@ -59,7 +59,7 @@ end
 # 2-3D Foward pass: Input X, Output Y
 function forward(X::AbstractArray{T, N}, AN::ActNorm; logdet=nothing) where {T, N}
     isnothing(logdet) ? logdet = (AN.logdet && ~AN.is_reversed) : logdet = logdet
-    inds = [i!=(N-1) ? 1 : (:) for i=1:N]
+    inds = [i!=(N-1) ? 1 : Colon() for i=1:N]
```
Comment (alisiahkoohi) [inline src/layers/invertible_layer_actnorm.jl:62]: Is this related to style or are there other benefits?

Comment (mloubout) [reply src/layers/invertible_layer_actnorm.jl:62]: Is easy to forget parenthesis and leads to unreadable error. `Colon` is safer IMO

File: src/layers/invertible_layer_conv1x1.jl [hunk 2]
```diff
@@ -105,22 +105,13 @@ function partial_derivative_outer(v::CuArray{T, 1}) where T
     return outer
 end
 
-function mat_tens_i(out::AbstractArray{T, 3}, Mat::AbstractArray{T, 2},
+
```
Comment (alisiahkoohi) [inline src/layers/invertible_layer_conv1x1.jl:108]: A comment here would be appreciated.

File: src/networks/invertible_network_conditional_hint.jl [hunk 3]
```diff
@@ -124,18 +126,20 @@ function backward(ΔZx::AbstractArray{T, N}, ΔZy::AbstractArray{T, N}, Zx::Abst
                 ΔZx_, ΔZy_, Δθcl, Zx_, Zy_, ∇logdetcl = CH.CL[j].backward(ΔZx, ΔZy, Zx, Zy; set_grad=set_grad)
                 ΔZx, Δθx, Zx, ∇logdetx = CH.AN_X[j].backward(ΔZx_, Zx_; set_grad=set_grad)
                 ΔZy, Δθy, Zy, ∇logdety = CH.AN_Y[j].backward(ΔZy_, Zy_; set_grad=set_grad)
-                ∇logdet = cat(∇logdetx, ∇logdety, ∇logdetcl, ∇logdet; dims=1)
+                prepend!(∇logdetANX, ∇logdetx);prepend!(∇logdetANY, ∇logdety);prepend!(∇logdetCL, ∇logdetcl)
```
Comment (alisiahkoohi) [inline src/networks/invertible_network_conditional_hint.jl:129]: Split to multiple lines for readability. Same below.

File: src/networks/invertible_network_conditional_hint_multiscale.jl [hunk 4]
```diff
@@ -186,13 +188,13 @@ function backward(ΔZx::AbstractArray{T, N}, ΔZy::AbstractArray{T, N}, Zx::Abst
                     ΔZx_, ΔZy_, Δθcl, Zx_, Zy_, ∇logdetcl = CH.CL[i, j].backward(ΔZx, ΔZy, Zx, Zy; set_grad=set_grad)
                     ΔZx, Δθx, Zx, ∇logdetx = CH.AN_X[i, j].backward(ΔZx_, Zx_; set_grad=set_grad)
                     ΔZy, Δθy, Zy, ∇logdety = CH.AN_Y[i, j].backward(ΔZy_, Zy_; set_grad=set_grad)
-                    ∇logdet = cat(∇logdetx, ∇logdety, ∇logdetcl, ∇logdet; dims=1)
+                    prepend!(∇logdetANX, ∇logdetx);prepend!(∇logdetANY, ∇logdety);prepend!(∇logdetCL, ∇logdetcl)
```
Comment (alisiahkoohi) [inline src/networks/invertible_network_conditional_hint_multiscale.jl:191]: Here too.

File: test/test_layers/test_layer_conv1x1.jl [hunk 5]
```diff
@@ -2,8 +2,9 @@
 # Author: Philipp Witte, pwitte3@gatech.edu
 # Date: January 2020
 
-using InvertibleNetworks, LinearAlgebra, Test
+using InvertibleNetworks, LinearAlgebra, Test, Flux, Random
```
Comment (alisiahkoohi) [inline test/test_layers/test_layer_conv1x1.jl:6]: maybe easier to toggle between cpu and gpu by defining `device = cpu` as default as assigning the necessarily variables to `|> device`.

**Review by alisiahkoohi — CHANGES_REQUESTED** [review]
Thanks for making these changes to improve the memory usage. Would it be helpful to have some tests that run on gpu that we run locally? For example to ensure to scalar operation happens on gpu.

**Comment by alisiahkoohi** [comment]
> > Would it be possible not to change the order of parameter retrieval? As you said, may cause issue older pretrained models.
> 
> Is kinda of a pain because right now it's done by hand on each network but there is not easy "generalized" way to get the order obtained by hand because depending on the network it makes and implicit assumption on the order in which the layers (struct's field) are used.

I agree that it looks much better now. Alternatively, would it be possible to create a similar function that does the operation in reverse? e.g., `put_params`?

**Comment by mloubout** [comment]
>e.g., put_params?

There is a `set_params!(network, theta)` already yes

**Review by alisiahkoohi — APPROVED** [review]
